File: leetcode_0046/cpp/leetcode_0046.cpp
//
// Created by wangzy on 2022/4/17.
//

#include<iostream>
#include<queue>
#include<vector>
#include <algorithm>
#include <unordered_map>
#include <unordered_set>
#include <list>
#include <deque>
#include <stack>
#include <sstream>
#include <typeinfo>
#include <climits>
#include <random>
#include <ctime>

using namespace std;

class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        //思路：回溯 dfs 每次交换当前元素与后序元素然后向后移动序列
        vector<vector<int>> res;
        vector<int> sub;
        dfs(res,nums,sub,0,nums.size()-1);
        return res;
    }
private:
    void dfs(vector<vector<int>>& res,vector<int>& nums,vector<int>& sub,int start,int end){
        if(start > end)
            return ;
        else if(start == end){
            sub.push_back(nums[start]);
            res.push_back(sub);
            sub.pop_back();
            return ;
        }
        for(int i = start;i <= end;i++){
            swap(nums[start],nums[i]);
            sub.push_back(nums[start]);
            dfs(res,nums,sub,start+1,end);
            sub.pop_back();
            swap(nums[start],nums[i]);
        }
    }
};


template<typename T>
string convertVecToStr(std::vector<T> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << ele;
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    return res;
}

template<typename T>
void printVecArr(std::vector<std::vector<T>> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << convertVecToStr(ele);
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    std::cout<<res<<std::endl;
}

template<typename T>
void printVec(std::vector<T> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << ele;
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    std::cout<<res<<std::endl;
}

int main(){
    vector<int> nums = {1,2,3};
    //nums = [1,2,3]
    auto res = Solution().permute(nums);
    printVecArr(res);
    //[[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]
}

File: leetcode_0046/java/leetcode_0046.java
import java.util.Arrays;
import java.util.LinkedList;
import java.util.List;

public class leetcode_0046 {
    class Solution {
        public List<List<Integer>> permute(int[] nums) {
            //思路：回溯 dfs 每次交换当前元素与后序元素然后向后移动序列
            List<List<Integer>> res = new LinkedList();
            LinkedList<Integer> sub = new LinkedList();
            dfs(res,nums,sub,0,nums.length - 1);
            return res;
        }
        private void dfs(List<List<Integer>> res,int[] nums,LinkedList<Integer> sub,int start,int end){
            if(start > end)
                return;
            else if(start == end){
                sub.addLast(new Integer(nums[start]));
                res.add(new LinkedList(sub));
                sub.removeLast();
                return;
            }
            for(int i = start;i <= end;i++){
                swap(nums,start,i);
                sub.addLast(new Integer(nums[start]));
                dfs(res,nums,sub,start+1,end);
                sub.removeLast();
                swap(nums,start,i);
            }
        }
        private void swap(int[] nums,int i,int j){
            int tmp = nums[i];
            nums[i] = nums[j];
            nums[j] = tmp;
        }
        public <T> void printListArr(List<List<T>> list) {
            for (int i = 0; i < list.size(); i++) {
                if (i == list.size() - 1) {
                    if (i == 0)
                        System.out.printf("[ ");
                    System.out.println(Arrays.toString(list.get(i).toArray(Integer[]::new)) + " ] ");
                } else {
                    if (i == 0)
                        System.out.printf("[ ");
                    System.out.printf(Arrays.toString(list.get(i).toArray(Integer[]::new)) + " , ");
                }
            }
        }
    }
    public static void main(String[] args) {
        leetcode_0046.Solution solution = new leetcode_0046().new Solution();
        int[] nums = {1,2,3};
        //nums = [1,2,3]
        List<List<Integer>> res = solution.permute(nums);
        solution.printListArr(res);
        //[[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]
    }
}
